src/fidesops/service/authentication/authentication_strategy_oauth2.py

```diff
@@ -0,0 +1,256 @@
+from fidesops.schemas.saas.saas_config import ClientConfig, QueryParam, SaaSRequest
+from fidesops.schemas.saas.strategy_configuration import (
+    OAuth2AuthenticationConfiguration,
+    StrategyConfiguration,
+)
+from fidesops.service.authentication.authentication_strategy import (
+    AuthenticationStrategy,
+)
+from fidesops.service.connectors.saas.authenticated_client import AuthenticatedClient
+from fidesops.util.logger import NotPii
+from fidesops.util.saas_util import assign_placeholders, map_param_values
+
+logger = logging.getLogger(__name__)
+
+
+class OAuth2AuthenticationStrategy(AuthenticationStrategy):
+    """
+    Checks the expiration date on the stored access token and refreshes
+    it if needed using the configured token refresh request.
+    """
+
+    strategy_name = "oauth2"
+
+    def __init__(self, configuration: OAuth2AuthenticationConfiguration):
+        self.expires_in = configuration.expires_in
+        self.authorization_request = configuration.authorization_request
+        self.token_request = configuration.token_request
+        self.refresh_request = configuration.refresh_request
+
+    def add_authentication(
+        self, request: PreparedRequest, connection_config: ConnectionConfig
+    ) -> PreparedRequest:
+        """
+        Checks the expiration date on the existing access token and refreshes if necessary.
+        The existing/updated access token is then added to the request as a bearer token.
+        """
+
+        access_token = connection_config.secrets.get("access_token")
+        if access_token is None:
+            raise FidesopsException(
+                f"OAuth2 access token not found for {connection_config.key}, please "
+                f"authenticate connection via /api/v1/connection/{connection_config.key}/authorize"
+            )
+
+        # The refresh request is optional since the OAuth2 spec does not require
+        # a refresh token to be returned as part of an access token request
+        #
+        # https://datatracker.ietf.org/doc/html/rfc6749#section-5.1
+        if self.refresh_request:
+            expires_at = int(connection_config.secrets.get("expires_at"))
+            if self._close_to_expiration(expires_at):
+                refresh_response = self._call_token_request(
+                    "refresh", self.refresh_request, connection_config
+                )
+                access_token = self._process_response(
+                    refresh_response, connection_config
+                )
+
+        # add access_token to request
+        request.headers["Authorization"] = "Bearer " + access_token
+        return request
+
+    @staticmethod
+    def _close_to_expiration(expires_at: int) -> bool:
+        """Check if the access_token will expire in the next 10 minutes"""
+        return expires_at < (datetime.utcnow() + timedelta(minutes=10)).timestamp()
+
+    @staticmethod
+    def _call_token_request(
+        action: str,
```

Comment:
Should `action` be restricted to being a certain set of values?

Reply:
Is this what you had in mind?
```
@staticmethod
def _call_token_request(
    action: Literal["access", "refresh"],
    token_request: SaaSRequest,
    connection_config: ConnectionConfig,
) -> Dict[str, Any]:
```

Reply:
that works well